parity-clib/src/lib.rs

```diff
@@ -33,132 +34,140 @@ pub struct ParityParams {
+	panic::catch_unwind(|| {
+		*output = ptr::null_mut();
+
+		let args = {
+			let arg_ptrs = slice::from_raw_parts(args, len);
+			let arg_lens = slice::from_raw_parts(args_lens, len);
 
-				args
-			};
-
-			match parity_ethereum::Configuration::parse_cli(&args) {
-				Ok(mut cfg) => {
-					// Always disable the auto-updater when used as a library.
-					cfg.args.arg_auto_update = "none".to_owned();
-
-					let cfg = Box::into_raw(Box::new(cfg));
-					*output = cfg as *mut _;
-					0
-				},
-				Err(_) => {
-					1
-				},
+			let mut args = Vec::with_capacity(len + 1);
+			args.push("parity".to_owned());
+
+			for (&arg, &len) in arg_ptrs.iter().zip(arg_lens.iter()) {
+				let string = slice::from_raw_parts(arg as *const u8, len);
+				match String::from_utf8(string.to_owned()) {
+					Ok(a) => args.push(a),
+					Err(_) => return 1,
+				};
 			}
-		}).unwrap_or(1)
-	}
+
+			args
+		};
+
+		match parity_ethereum::Configuration::parse_cli(&args) {
+			Ok(mut cfg) => {
+				// Always disable the auto-updater when used as a library.
+				cfg.args.arg_auto_update = "none".to_owned();
+
+				let cfg = Box::into_raw(Box::new(cfg));
+				*output = cfg as *mut _;
+				0
+			},
+			Err(_) => {
+				1
+			},
+		}
+	}).unwrap_or(1)
 }
 
 #[no_mangle]
-pub extern fn parity_config_destroy(cfg: *mut c_void) {
-	unsafe {
-		let _ = panic::catch_unwind(|| {
-			let _cfg = Box::from_raw(cfg as *mut parity_ethereum::Configuration);
-		});
-	}
+pub unsafe extern fn parity_config_destroy(cfg: *mut c_void) {
+	let _ = panic::catch_unwind(|| {
+		let _cfg = Box::from_raw(cfg as *mut parity_ethereum::Configuration);
+	});
 }
 
 #[no_mangle]
-pub extern fn parity_start(cfg: *const ParityParams, output: *mut *mut c_void) -> c_int {
-	unsafe {
-		panic::catch_unwind(|| {
-			*output = ptr::null_mut();
-			let cfg: &ParityParams = &*cfg;
-
-			let config = Box::from_raw(cfg.configuration as *mut parity_ethereum::Configuration);
-
-			let on_client_restart_cb = {
-				struct Cb(Option<extern "C" fn(*mut c_void, *const c_char, usize)>, *mut c_void);
-				unsafe impl Send for Cb {}
-				unsafe impl Sync for Cb {}
-				impl Cb {
-					fn call(&self, new_chain: String) {
-						if let Some(ref cb) = self.0 {
-							cb(self.1, new_chain.as_bytes().as_ptr() as *const _, new_chain.len())
-						}
+pub unsafe extern fn parity_start(cfg: *const ParityParams, output: *mut *mut c_void) -> c_int {
+	panic::catch_unwind(|| {
+		*output = ptr::null_mut();
+		let cfg: &ParityParams = &*cfg;
+
+		let config = Box::from_raw(cfg.configuration as *mut parity_ethereum::Configuration);
+
+		let on_client_restart_cb = {
+			struct Cb(Option<extern "C" fn(*mut c_void, *const c_char, usize)>, *mut c_void);
+			unsafe impl Send for Cb {}
+			unsafe impl Sync for Cb {}
+			impl Cb {
+				fn call(&self, new_chain: String) {
+					if let Some(ref cb) = self.0 {
+						cb(self.1, new_chain.as_bytes().as_ptr() as *const _, new_chain.len())
 					}
 				}
```

Comment:
"Duplicate struct definition", the only thing the differs is that this one is inside Option and https://github.com/paritytech/parity-ethereum/blob/4ce97f9554d2353f225d8f47f615e85a019e0316/parity-clib/src/lib.rs#L159-#L167 is not!

Can you move the `"local struct definitions"` to one `file scope definition` and rename it to `Callback`? I think the full name is worth the effort here!